### PR TITLE
LPS-124200 Refactor useChartState to be able to use it within main and detail views

### PR DESCRIPTION
**Pull request by dgarciasarai**
**Description**

Decouple the `chartState` from `Chart` component to make it reusable in other components/detail views (referral and social).

On top of https://github.com/liferay-frontend/liferay-portal/pull/573.

**Solution**

- Create a new `ChartStateContextProvider` and inject it in the entry point
- Create custom hooks for every old action
- Create a custom hook for the date title
- Create a custom hook for the previous period button (if it's disabled or not)
- Clean all unnecessary parameters in each component

**Note**
This pull does not change any UI element, it's a code refactor.

### CODE REVIEW

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
Hey @liferay-frontend, can you give me feedback? This pull is on top of https://github.com/liferay-frontend/liferay-portal/pull/578 and it's related so @liferay-tango needs it as soon as possible. Thanks!

**Comment by dgarciasarai** [comment]
ci:test:relevant

**Comment by dgarciasarai** [comment]
ci:test:relevant

**Comment by dgarciasarai** [comment]
I rebased the changes and resolved the conflicts. Thanks!

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 1]
```diff
@@ -140,17 +147,36 @@ export default function Chart({
...
+
+	const setLoading = useSetLoading();
+
+	const changeTimeSpanKey = useChangeTimeSpanKey();
+
+	const previousTimeSpan = usePreviousTimeSpan();
+
+	const nextTimeSpan = useNextTimeSpan();
+
+	const isPreviousPeriodButtonDisabled = useIsPreviousPeriodButtonDisabled();
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:162]: Cosmetic thing: Brian might ask for these to be sorted (although he may not know that they are side-effect-free, so perhaps he'll say nothing).

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 2]
```diff
@@ -140,17 +147,36 @@ export default function Chart({
...
+
+	const dateFormatters = useMemo(() => dateFormat(languageTag), [
+		languageTag,
+	]);
+
+	const {firstDate, lastDate} = useDateTitle();
+
+	const title = useMemo(() => {
+		return dateFormatters.formatChartTitle([firstDate, lastDate]);
+	}, [dateFormatters, firstDate, lastDate]);
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:172]: Just a due diligience question: are you sure the computational cost that you're saving with `useMemo` here is greater than what you're spending? (I don't know how expensive `dateFormat()` and `formatChartTitle()` are.)

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js [hunk 3]
```diff
@@ -9,44 +9,134 @@
...
+			const firstDateLabel = histogram[0].label;
+
+			const firstDate = new Date(firstDateLabel);
+			const publishedDate = new Date(publishDate);
+
+			return firstDate < publishedDate;
+		}
+
+		return true;
+	}, [histogram, publishDate]);
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js:121]: This is an example of a `useMemo` call that doesn't look worth it. The contents will be super fast to compute and create very little garbage (two `Date` objects), so I wouldn't expect this to have any measurable effect.

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 4]
```diff
@@ -139,14 +146,24 @@ export default function Chart({
...
+
+	const addDataSetItems = useAddDataSetItems();
+
+	const setLoading = useSetLoading();
+
+	const changeTimeSpanKey = useChangeTimeSpanKey();
+
+	const previousTimeSpan = usePreviousTimeSpan();
+
+	const nextTimeSpan = useNextTimeSpan();
```
Comment (p2kmgcl) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:160]: why these new lines between declarations? (this is a really useful comment)

File: modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js [hunk 5]
```diff
@@ -85,20 +86,24 @@ describe('Navigation', () => {
...
-					timeSpanOptions={mockTimeSpanOptions}
-					viewURLs={mockViewURLs}
-				/>
+				>
+					<Navigation
+						author={testProps.author}
+						canonicalURL={testProps.canonicalURL}
+						endpoints={mockEndpoints}
+						languageTag={testProps.languageTag}
+						onSelectedLanguageClick={() => {}}
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js:99]: One thing to bear in mind when passing in a no-op callback is that if the component you're rendering uses that prop in a `useEffect` or `useMemo` etc, it will "see" a new callback every single render, and that can lead to additional re-renders. One way to avoid this, or make your code robust against that possibly happening in the future is to hoist out a `noop` function to the top level of this module and always pass that in instead. You can use it in all the places you're passing a no-op function.

```js
const noop = () => {};
```

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js [hunk 6]
```diff
@@ -49,25 +49,46 @@ export const ChartStateContextProvider = ({
...
-				type: ADD_DATA_SET_ITEMS,
-				validAnalyticsConnection,
-			}),
-		changeTimeSpanKey: (payload) =>
-			dispatch({payload, type: CHANGE_TIME_SPAN_KEY}),
-		nextTimeSpan: () => dispatch({type: NEXT_TIME_SPAN}),
-		previousTimeSpan: () => dispatch({type: PREV_TIME_SPAN}),
-		setLoading: () => dispatch({type: SET_LOADING}),
-	});
+	return (payload) =>
```
Comment (p2kmgcl) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js:59]: If you're using these functions inside useEffect calls you should wrap them with a useCallback, otherwise the useEffect will be triggered on every component call.

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 7]
```diff
@@ -25,9 +25,18 @@ import {
...
 
+import {
+	useAddDataSetItems,
+	useChangeTimeSpanKey,
+	useChartState,
+	useDateTitle,
+	useIsPreviousPeriodButtonDisabled,
+	useNextTimeSpan,
+	usePreviousTimeSpan,
+	useSetLoading,
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:36]: Interesting choice to structure this using so many hooks. My general rule of thumb is that hooks are all about _reuse_, so if a hook is only used in exactly one place, I wouldn't make it a custom hook. It seems that most of these are once-only hooks.

I think it's fine if you want to keep it this way, so just planting this as food for thought. The real trade-off here is because the hook custom hook is a layer of indirection, of abstraction, that conceals what's really happening. If "what's really happening" is verbose and painful to duplicate, the hook provides us with a way to encapsulate the functionality in a small unit (ie. reuse). If "what's really happening" is complex and difficult to understand, the hook can make the system more intellectually manageable, although for this to be true the concepts and naming have to be clear, otherwise it might be a lateral step at best. Conversely, if it is _not_ complex and difficult to understand, then the indirection might end up obfuscating the structure and relationships in the data model, and that could be information that would be useful to have in people's mental models.

In other words, as always in engineering, a set of trade-offs. But as a general starting point, the "hooks are for reuse" rule of thumb has served me pretty well.

**Review by wincent — APPROVED** [review]
Looks good to me.

Some comments to think about, but I leave up to you what you change or don't change. No blockers.

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 4 above)
Comment (wincent) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:160]: Because [Brian wants it that way](https://github.com/liferay/liferay-frontend-projects/issues/15).

Comment (p2kmgcl) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:160]: :open_mouth: I thought that was only if they are not sorted

Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:160]: Yes, that is the reason! As Greg suggested in his comment, I'm going to reorder alphabetically these declarations as much as possible :)

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 1 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:162]: I separated between variables and functions and sort each group. Let's see.

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js (hunk 6 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js:59]: Ok, I'll add useCallback in those that I used inside useEffect calls 👍  Thanks!

Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js:59]: In this case I will change `addDataSetItems` and `setLoading` (`Chart` component).

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 2 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:172]: I think maybe I can create a variable just calling `dateFormatters.formatCharTitle` without the memoization. `dateFormat` is just the initialization of a set of methods to transform data with a specific format and `formatChartTitle` is inside that methods and inside, it uses `Intl.DateTimeFormat` with specific properties to represent it numeric, string, etc. I think it's not super expensive... 🤔 Also, I create a new hook to get the `firstDate` and `lastDate` to calculate the chart title (see `useDateTitle). So maybe we can remove this `useMemo`. What do you think?

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js (hunk 3 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js:121]: This `useMemo` was used in Chart component with the date formatter call. I just move the code. So I can remove the `useMemo`, right?

File: modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js (hunk 5 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js:99]: Should I replace all occurrences for "() => {}" with "noop" and set it as you said outside the component? 🤔 Even in the tests? What is the magic here? How React know that we should not re-render things? Just to understand because this is a new concept to me :)

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 7 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:36]: I totally agree with you and that's the principal idea of this pull request. 

I created those custom hooks because the idea is to make it reusable in almost every detail view and also decouple from the `Chart` component. We are actually working on Referral (see https://github.com/liferay-frontend/liferay-portal/pull/573, you can give feedback also 😛) and Social detail view (still WIP) from traffic channels. In those views, we have the same selector (7 Days, 30 Days) as in the first level and we want to share the state between those screens (main-detail).

Looking at that list once again, I think that the only one that maybe will be used in one component is the `useAddDataSetItems` (do you remember it right? This method converts the data to an understandable one for the chart, you helped me with this 😄 ). The others, as I said, will be used in detail view.

**Comment by dgarciasarai** [comment]
Thanks @wincent and @p2kmgcl for the feedback. I'm going to apply it and then forward the changes :)

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 7 above)
Comment (wincent) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:36]: So here's the other axis to think about, beyond re-use: if the same set of N hooks are used in multiple places, maybe they should just be one hook (or at least `< N` hooks); in this case, a few of them are just wrappers around `useContext`, that return `dispatch`, so "the one hook" might just be `useContext`. I'm not saying you should change this, but just wanted to plant it as a seed of an idea: that if you need to use a dozen custom hooks many times over to get something done, you may not have actually made the system easier to use overall (especially the cases where the hooks are internally simple: because the indirection just means that you have to jump to another place in order to see what they actually do, and you also have to remember what they all are etc).

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js (hunk 3 above)
Comment (wincent) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js:121]: I'd think that you could.

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
ci:test:relevant

**Comment by dgarciasarai** [comment]
Hey @wincent @p2kmgcl, I think I applied all your feedback, let me know if I messed up something before forwarding it. Thanks

File: modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js (hunk 5 above)
Comment (wincent) [reply modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js]: Well, there's no magic. The point is just that _any_ prop you pass into a component could end up causing that component to re-render unnecessarily _if_ that prop is not stable over time, _and_ is used as an input into a `useEffect` dependencies array, or `useCallback`, or `useMemo` etc. This may be subtle too, because if the prop is passed down to a grandchild or a great-grandchild, the place where the prop is causing a re-render may be buried several levels down.

So, the basic pattern is, _whenever_ you see `<Component prop={value} />`, two of the questions you should have in mind if you want to maximize interoperability are: "_Can_ `value` be stable over time? If so, _what_ do I have to do to make it stable?" If you do that as a matter of standard practice, your code is more likely to be performant now and in the future (when it gets transplanted into some unforeseen context).

In this case, `() => {}` is not stable over time, because every time you render, you make a new anonymous function and `(() => {}) !== (() => {})`. But given `const noop = () => {}` at the top-level of the module, `noop === noop` and is stable, so it's a good pattern to follow in general. (You should think about similar gotchas with default values for function parameters: if you have a param `callback` with a default — `callback = () => {}`, then _that's_ going to be a different fallback callback every time somebody doesn't pass a default parameter.)

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/StoreContext.js [hunk 8]
```diff
@@ -19,7 +19,9 @@ const INITIAL_STATE = {
 const ADD_HISTORICAL_WARNING = 'add-historical-warning';
 const ADD_WARNING = 'add-warning';
 
-export const StoreContext = createContext([INITIAL_STATE, () => {}]);
+const noop = () => {};
+
+export const StoreContext = createContext([INITIAL_STATE, noop]);
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/StoreContext.js:24]: In this specific instance you don't benefit from the `noop` trick because `createContext` is called exactly once.

File: modules/dxp/apps/analytics/analytics-reports-web/test/js/components/TrafficSources.js [hunk 9]
```diff
@@ -15,6 +15,8 @@ import React from 'react';
 
 import TrafficSources from '../../../src/main/resources/META-INF/resources/js/components/TrafficSources';
 
+const noop = () => {};
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/test/js/components/TrafficSources.js:18]: I also wouldn't worry about it in tests, unless you think it helps readability (which is the most important thing in tests anyway; you don't have to worry about re-renders in the context of a test suite because most `it` blocks render exactly once).

**Comment by wincent** [comment]
> Thanks @wincent and @p2kmgcl for the feedback. I'm going to apply it and then forward the changes :)

Changes all look fine to me.

File: modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js (hunk 5 above)
Comment (p2kmgcl) [reply modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js]: You can also reuse $.noop if you enable jQuery in portal, much better

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 7 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:36]: I'll take into account for next implementations. Thanks :)

File: modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js (hunk 5 above)
Comment (wincent) [reply modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js]: not-amused

File: modules/dxp/apps/analytics/analytics-reports-web/test/js/components/TrafficSources.js (hunk 9 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/test/js/components/TrafficSources.js:18]: Well, I replace all occurrences... OCD. My fault.

File: modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js (hunk 5 above)
Comment (p2kmgcl) [reply modules/dxp/apps/analytics/analytics-reports-web/test/js/components/Navigation.js]: :crying_cat_face:

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 7 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:36]: See the use of these custom hooks in ReferralDetail: https://github.com/liferay-frontend/liferay-portal/pull/582 (2 commits). Sooner or later it seems that I will have a unique custom hook for these `N` actions 🤔

**Comment by dgarciasarai** [comment]
ci:forward
